torchci/rockset/commons/__sql/flaky_workflows_jobs.sql

```diff
@@ -1,55 +1,65 @@
-with repeats as (
-    select
-        array_agg(j.id) as ids
-    from
-        workflow_job j
-        join workflow_run w on w.id = j.run_id
-    where
-        j._event_time >= PARSE_DATETIME_ISO8601(:startTime)
-        and j._event_time < PARSE_DATETIME_ISO8601(:stopTime)
-        and w.head_repository.full_name = :repo
-        and w.head_branch = :branch
-        AND w.event != 'workflow_run'
-        AND w.event != 'repository_dispatch'
-    group by
-        j.head_sha,
-        j.name,
-        w.name
-    having
-        count(*) > :count
-        and bool_or(
-            j.conclusion in ('failure', 'cancelled', 'time_out')
-        )
-),
-ids as (
-    select
-        ids.id
-    from
-        repeats,
-        unnest(repeats.ids as id) as ids
-)
-select
-    job.head_sha as sha,
-    CONCAT(w.name, ' / ', job.name) as jobName,
+WITH failed_jobs AS (
+  SELECT
+    FIRST_VALUE(job.conclusion) OVER(
+      PARTITION BY CONCAT(w.name, ' / ', job.name)
+      ORDER BY
+        push.head_commit.timestamp ROWS BETWEEN 1 PRECEDING
+        AND 1 FOLLOWING
+    ) = 'success'
+    AND NTH_VALUE(job.conclusion, 2) OVER(
+      PARTITION BY CONCAT(w.name, ' / ', job.name)
+      ORDER BY
+        push.head_commit.timestamp ROWS BETWEEN 1 PRECEDING
+        AND 1 FOLLOWING
+    ) = 'failure'
+    AND LAST_VALUE(job.conclusion) OVER(
+      PARTITION BY CONCAT(w.name, ' / ', job.name)
+      ORDER BY
+        push.head_commit.timestamp ROWS BETWEEN 1 PRECEDING
+        AND 1 FOLLOWING
+    ) = 'success' AS flaky,
```

Comment:
I suspect I'm misunderstanding something here.

It seems like when a job fails, retryBot is querying rockset to see if that particular job's run failed for a flaky reason, and it does this by checking the previous and next commits tested.

However, retrybot receives the failure right when the workflow ends. At that point in time, the next job would not have had a chance to complete running, so this query can never return a positive response when queried by retry bot.

Am I perhaps missing something?

Reply:
Good point!  I think you're right.  The query is fine as it returns the correct flaky job.  But there is the edge case where there is no newer job or the newer job hasn't finished yet.  It means that there is not enough data to decide.  I think I could tweak the logic a bit to retry the job that is immediately before instead.  Basically, it would be `Green (Retry the previous flaky red job) Red Green` instead of `Green Red (Retry itself) Green`

Reply:
The correct logic is now in place.  Essentially, the bot would also retry previous job if the current job succeeds while the previous one failed (and the one before that again succeeded, thus still Green/Red/Green).  On the bot:

```
await retryPreviousWorkflow(
  ctx,
  owner,
  repo,
  workflowName,
  workflowJobs.filter((job) =>
    SUCCESS_CONCLUSIONS.includes(job.conclusion!)
  )
);
```

The fun part is in the Rockset query:

```
FIRST_VALUE(job.conclusion) OVER(
  PARTITION BY CONCAT(w.name, ' / ', job.name)
  ORDER BY
    push.head_commit.timestamp DESC ROWS BETWEEN CURRENT ROW
    AND 2 FOLLOWING
) = 'success'
AND NTH_VALUE(job.conclusion, 2) OVER(
  PARTITION BY CONCAT(w.name, ' / ', job.name)
  ORDER BY
    push.head_commit.timestamp DESC ROWS BETWEEN CURRENT ROW
    AND 2 FOLLOWING
) = 'failure'
AND LAST_VALUE(job.conclusion) OVER(
  PARTITION BY CONCAT(w.name, ' / ', job.name)
  ORDER BY
    push.head_commit.timestamp DESC ROWS BETWEEN CURRENT ROW
    AND 2 FOLLOWING
) = 'success' AS prev_flaky,
```

As you can see, this is now a windows of the current row and 2 rows before it.  I try it out on Rockset and it works pretty well because a failed job usually (if not always) finished before successful jobs.  

There is still a small edge case where older job could take longer to finish (stuck or hang for example), but that should be ok IMO as we can always retry manually from time to time.  In fact, that case would also apply to the previous approach of checking the jobs before and after.